### PR TITLE
common: Add stateful thread worker and unique function utilities

### PR DESCRIPTION
**Pull request by ReinUsesLisp**
- Add a unique function utility. Unlike std::function, this one only requires for the captured objects to be movable. Copies are not supported.
- Improve thread worker utility. Add support for "stateful" threads, waiting for tasks to complete and interrupting waits through `std::stop_token`.
- Change the thread worker utility to use a `Common::UniqueFunction` instead of `std::function`.

Credits to @FernandoS27 for fixing some data races on this abstraction.

### CODE REVIEW

File: src/common/thread_worker.h [hunk 1]
```diff
@@ -7,24 +7,110 @@
...
 private:
-    std::vector<std::thread> threads;
-    std::queue<std::function<void()>> requests;
+    std::queue<Task> requests;
     std::mutex queue_mutex;
-    std::condition_variable condition;
-    std::atomic_bool stop{};
+    std::condition_variable_any condition;
+    std::condition_variable wait_condition;
+    std::atomic<size_t> work_scherduled{};
```
Comment (ameerj) [inline src/common/thread_worker.h:106]:
```suggestion
    std::atomic<size_t> work_scheduled{};
```

File: src/common/unique_function.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+#pragma once
+
+#include <memory>
+#include <utility>
+
+namespace Common {
+
+template <typename ResultType, typename... Args>
+class UniqueFunction {
```
Comment (lioncash) [inline src/common/unique_function.h:13]: This class should probably have some documentation explaining what it is and what its purpose is.

File: src/common/unique_function.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    };
+
+public:
+    UniqueFunction() = default;
+
+    template <typename Functor>
+    UniqueFunction(Functor&& functor)
+        : callable{std::make_unique<Callable<Functor>>(std::move(functor))} {}
+
+    UniqueFunction& operator=(UniqueFunction<ResultType, Args...>&& rhs) noexcept {
```
Comment (lioncash) [inline src/common/unique_function.h:41]: ....................
```suggestion
    UniqueFunction& operator=(UniqueFunction&& rhs) noexcept {
```

File: src/common/unique_function.h [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        virtual ResultType operator()(Args...) = 0;
+    };
+
+    template <typename Functor>
+    class Callable final : public CallableBase {
+    public:
+        Callable(Functor&& functor_) : functor{std::move(functor_)} {}
+        ~Callable() override = default;
+
+        ResultType operator()(Args... args) override {
```
Comment (lioncash) [inline src/common/unique_function.h:26]:
```suggestion
        ResultType operator()(Args&&... args) override {
```

File: src/common/unique_function.h [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+    UniqueFunction& operator=(UniqueFunction<ResultType, Args...>&& rhs) noexcept {
+        callable = std::move(rhs.callable);
+        return *this;
+    }
+
+    UniqueFunction(UniqueFunction<ResultType, Args...>&& rhs) noexcept
+        : callable{std::move(rhs.callable)} {}
+
+    ResultType operator()(Args... args) const {
```
Comment (lioncash) [inline src/common/unique_function.h:49]:
```suggestion
    ResultType operator()(Args&&... args) const {
```

File: src/common/unique_function.h [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    template <typename Functor>
+    UniqueFunction(Functor&& functor)
+        : callable{std::make_unique<Callable<Functor>>(std::move(functor))} {}
+
+    UniqueFunction& operator=(UniqueFunction<ResultType, Args...>&& rhs) noexcept {
+        callable = std::move(rhs.callable);
+        return *this;
+    }
+
+    UniqueFunction(UniqueFunction<ResultType, Args...>&& rhs) noexcept
```
Comment (lioncash) [inline src/common/unique_function.h:46]:
```suggestion
    UniqueFunction(UniqueFunction&& rhs) noexcept
```

File: src/common/unique_function.h [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+    ResultType operator()(Args... args) const {
+        return (*callable)(std::forward<Args>(args)...);
+    }
+
+    explicit operator bool() const noexcept {
+        return callable != nullptr;
+    }
+
+    UniqueFunction& operator=(const UniqueFunction<ResultType, Args...>&) = delete;
```
Comment (lioncash) [inline src/common/unique_function.h:57]:
```suggestion
    UniqueFunction& operator=(const UniqueFunction&) = delete;
```

File: src/common/unique_function.h [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    UniqueFunction(const UniqueFunction<ResultType, Args...>&) = delete;
```
Comment (lioncash) [inline src/common/unique_function.h:58]:
```suggestion
    UniqueFunction(const UniqueFunction&) = delete;
```

File: src/common/unique_function.h [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+#include <utility>
+
+namespace Common {
+
+template <typename ResultType, typename... Args>
+class UniqueFunction {
+    class CallableBase {
+    public:
+        virtual ~CallableBase() = default;
+        virtual ResultType operator()(Args...) = 0;
```
Comment (lioncash) [inline src/common/unique_function.h:17]:
```suggestion
        virtual ResultType operator()(Args&&...) = 0;
```

**Comment by ReinUsesLisp** [comment]
Addressed comments except for documenting UniqueFunction. I'll do that soon.

**Review by ameerj — APPROVED** [review]

**Pull request merged**